### PR TITLE
[CIS-309] Add channel reads

### PR DESCRIPTION
**Pull request by b-onc**
With this PR, I'm introducing a big terminology change: MessageRead -> ChannelRead

Why?

Because, the `read` objects returned with channel query are actually channel reads. They don't point to a single message, they signify when was the last time a user has read the channel.

Storing these objects in DB IMHO makes sense since we'd like to display read state on messages even when offline.

I didn't add any calculation about unread counts or if a message is read in this PR to keep it small.

### CODE REVIEW

File: Sources_v3/APIClient/Endpoints/Payloads/UserPayloads.swift [hunk 1]
```diff
@@ -7,7 +7,7 @@ import Foundation
 enum UserPayloadsCodingKeys: String, CodingKey {
     case id
     case role
-    case name
+    case name // TODO: remove name and avatarURL below, since they're parsed as extraData
```
Comment (VojtaStavik) [inline Sources_v3/APIClient/Endpoints/Payloads/UserPayloads.swift:10]: Wouldn't it be easier to just delete them instead of this todo?

File: Sources_v3/Database/DTOs/ChannelReadDTO.swift [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        return dto
+    }
+}
+
+extension ChannelReadModel {
+    static func create(fromDTO dto: ChannelReadDTO) -> ChannelReadModel {
+        let user = UserModel<ExtraData.User>.create(fromDTO: dto.user)
+        return .init(lastReadAt: dto.lastReadAt, unreadMessagesCount: dto.unreadMessageCount, user: user)
+    }
+}
```
Comment (VojtaStavik) [inline Sources_v3/Database/DTOs/ChannelReadDTO.swift:66]: Can you expose this functionality using `asModel()` like we have it for other DTOs? (but Channel - no one fixed it there, yet)

File: Sources_v3/Database/DTOs/ChannelDTO.swift [hunk 3]
```diff
@@ -160,6 +163,8 @@ extension ChannelModel {
             .load(for: dto.cid, limit: 25, context: dto.managedObjectContext!)
             .map { $0.asModel() }
         
+        let reads = dto.reads.map(ChannelReadModel<ExtraData>.create(fromDTO:))
+
```
Comment (VojtaStavik) [inline Sources_v3/Database/DTOs/ChannelDTO.swift:167]: Ideally, you'd do just `dto.reads.map { $0.asModel() }` like above with the messages.

File: Sources_v3/Database/DTOs/UserDTO.swift [hunk 4]
```diff
@@ -33,7 +33,7 @@ extension UserDTO {
     ///   - id: The id of the user to fetch
     ///   - context: The context used to fetch `UserDTO`
     ///
-    fileprivate static func load(id: String, context: NSManagedObjectContext) -> UserDTO? {
+    static func load(id: String, context: NSManagedObjectContext) -> UserDTO? {
```
Comment (VojtaStavik) [inline Sources_v3/Database/DTOs/UserDTO.swift:36]: Why this change? Isn't `context.user(id:)` what you need?

File: Sources_v3/Database/DTOs/UserDTO.swift [hunk 5]
```diff
@@ -102,7 +102,7 @@ extension UserDTO {
 }
 
 extension UserModel {
-    fileprivate static func create(fromDTO dto: UserDTO) -> UserModel {
+    static func create(fromDTO dto: UserDTO) -> UserModel {
```
Comment (VojtaStavik) [inline Sources_v3/Database/DTOs/UserDTO.swift:105]: Please use `asModel()` for this and keep this private

File: Sources_v3/Models/ChannelReadModel.swift [hunk 6]
```diff
@@ -0,0 +1,24 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+public struct ChannelReadModel<ExtraData: ExtraDataTypes> {
```
Comment (VojtaStavik) [inline Sources_v3/Models/ChannelReadModel.swift:7]: Documentation? ;-)

File: Sources_v3/APIClient/Endpoints/ChannelEndpoints.swift [hunk 7]
```diff
@@ -107,4 +107,14 @@ extension Endpoint {
...
+    
+    static func markRead(cid: ChannelId) -> Endpoint<EmptyResponse> {
+        .init(
+            path: "channels/\(cid.type)/\(cid.id)/read",
+            method: .post,
+            queryItems: nil,
+            requiresConnectionId: false,
+            body: nil
+        )
+    }
```
Comment (VojtaStavik) [inline Sources_v3/APIClient/Endpoints/ChannelEndpoints.swift:119]: This should be tested

File: Sources_v3/Controllers/ChannelController/ChannelController.swift [hunk 8]
```diff
@@ -575,6 +575,17 @@ public extension ChannelControllerGeneric {
...
+    /// Marks the channel as read.
+    /// - Parameter completion: The completion. Will be called on a **callbackQueue** when the network request is finished.
+    ///                         If request fails, the completion will be called with an error.
+    func markRead(completion: ((Error?) -> Void)? = nil) {
+        worker.markRead(cid: channelId) { [weak self] error in
+            self?.callback {
+                completion?(error)
+            }
+        }
+    }
```
Comment (VojtaStavik) [inline Sources_v3/Controllers/ChannelController/ChannelController.swift:588]: This should be tested. You also need to guard for `isChannelAlreadyCreated` because it's not possible to mark read channel that doesn't exist yet.

File: Sources_v3/APIClient/Endpoints/ChannelEndpoints.swift [hunk 9]
```diff
@@ -117,4 +117,14 @@ extension Endpoint {
...
+    
+    static func markAllRead() -> Endpoint<EmptyResponse> {
+        .init(
+            path: "channels/read",
+            method: .post,
+            queryItems: nil,
+            requiresConnectionId: false,
+            body: nil
+        )
+    }
```
Comment (VojtaStavik) [inline Sources_v3/APIClient/Endpoints/ChannelEndpoints.swift:129]: missing tests

File: Sources_v3/Controllers/ChannelListController/ChannelListController.swift [hunk 10]
```diff
@@ -157,6 +157,16 @@ public extension ChannelListControllerGeneric {
...
+    
+    /// Marks all channels for a user as read.
+    /// - Parameter completion: Called when the API call is finished. Called with `Error` if the remote update fails.
+    func markAllRead(completion: ((Error?) -> Void)? = nil) {
+        worker.markAllRead { [weak self] error in
+            self?.callback {
+                completion?(error)
+            }
+        }
+    }
```
Comment (VojtaStavik) [inline Sources_v3/Controllers/ChannelListController/ChannelListController.swift:169]: missing tests + guard for new channels

File: Sources_v3/Controllers/ChannelListController/ChannelListController_Tests.swift [hunk 11]
```diff
@@ -58,7 +58,7 @@ class ChannelListController_Tests: StressTestCase {
         XCTAssertEqual(controller.state, .localDataFetched)
         
         // Simulate successfull network call.
-        env.channelQueryUpdater?.update_completion?(nil)
+        env.channelListUpdater?.update_completion?(nil)
```
Comment (VojtaStavik) [inline Sources_v3/Controllers/ChannelListController/ChannelListController_Tests.swift:61]: This is probably not correct?? Or I don't understand what's happening here in the "Add markAllRead functionality and tests" commit

**Review by VojtaStavik — CHANGES_REQUESTED** [review]
What happened to our test coverage?? 😅 Otherwise looks like, thank you! I like the name change btw.

File: Sources_v3/APIClient/Endpoints/Payloads/UserPayloads.swift (hunk 1 above)
Comment (b-onc) [reply Sources_v3/APIClient/Endpoints/Payloads/UserPayloads.swift:10]: Oh yeah, that TODO was actually for myself to not forget 😅

**Comment by VojtaStavik** [comment]
@b-onc I think you can add the unread message count in the PR, too. It's simply the number of messages newer than the current user's "read" :)

File: Sources_v3/Database/DTOs/ChannelReadDTO.swift (hunk 2 above)
Comment (b-onc) [reply Sources_v3/Database/DTOs/ChannelReadDTO.swift:66]: I shall fix both!

File: Sources_v3/Database/DTOs/UserDTO.swift (hunk 4 above)
Comment (b-onc) [reply Sources_v3/Database/DTOs/UserDTO.swift:36]:
```swift
func user(id: UserId) -> UserDTO? {
        UserDTO.load(id: id, context: self)
    }
```
I need `loadOrCreate`

Comment (b-onc) [reply Sources_v3/Database/DTOs/UserDTO.swift:36]: Ooh you mean I didn't need `UserDTO.load`.. right, i'll restore `fileprivate`

File: Sources_v3/APIClient/Endpoints/Payloads/UserPayloads.swift (hunk 1 above)
Comment (VojtaStavik) [reply Sources_v3/APIClient/Endpoints/Payloads/UserPayloads.swift:10]: Glad it worked out

File: Sources_v3/Controllers/ChannelListController/ChannelListController_Tests.swift (hunk 11 above)
Comment (b-onc) [reply Sources_v3/Controllers/ChannelListController/ChannelListController_Tests.swift:61]: I've mixed the renaming commit with add markAllRead commit... Since I renamed `ChannelListQueryUpdater -> ChannelListUpdater` I've also renamed related properties in tests

Comment (VojtaStavik) [reply Sources_v3/Controllers/ChannelListController/ChannelListController_Tests.swift:61]: I got confused because the original naming here was clearly wrong, too. Sounds like the "channel query" updater, not "channel list query"

File: Sources_v3/Database/DTOs/ChannelReadDTO.swift (hunk 2 above)
Comment (VojtaStavik) [reply Sources_v3/Database/DTOs/ChannelReadDTO.swift:66]: Please don't, it's a huge change. That's why no one done it yet. It's a topic for different PR

Comment (VojtaStavik) [reply Sources_v3/Database/DTOs/ChannelReadDTO.swift:66]: @b-onc It's related to introducing `ChannelDatabaseSession` and `MemberDatabaseSession` and other changes. It's a tech-debt issue -> CIS-332

File: Sources_v3/Controllers/ChannelController/ChannelController.swift (hunk 8 above)
Comment (b-onc) [reply Sources_v3/Controllers/ChannelController/ChannelController.swift:588]: We should leave the guarding part out IMHO. Backend error response will be plenty enough to debug, and we'd like to prevent failures on `client.channelController(cid: cid).markRead()`

File: Sources_v3/APIClient/Endpoints/Payloads/UserPayloads.swift (hunk 1 above)
Comment (b-onc) [reply Sources_v3/APIClient/Endpoints/Payloads/UserPayloads.swift:10]: Fixed in 66b7f4c514244952de084ec846f05d090f03f866

File: Sources_v3/Database/DTOs/ChannelReadDTO.swift (hunk 2 above)
Comment (b-onc) [reply Sources_v3/Database/DTOs/ChannelReadDTO.swift:66]: Fixed in 62b46eb0b0bef360c33b591d1dbceb93446701e2

File: Sources_v3/Database/DTOs/ChannelDTO.swift (hunk 3 above)
Comment (b-onc) [reply Sources_v3/Database/DTOs/ChannelDTO.swift:167]: Fixed in 62b46eb0b0bef360c33b591d1dbceb93446701e2

File: Sources_v3/Database/DTOs/UserDTO.swift (hunk 4 above)
Comment (b-onc) [reply Sources_v3/Database/DTOs/UserDTO.swift:36]: Fixed in 62b46eb0b0bef360c33b591d1dbceb93446701e2

File: Sources_v3/Database/DTOs/UserDTO.swift (hunk 5 above)
Comment (b-onc) [reply Sources_v3/Database/DTOs/UserDTO.swift:105]: Fixed in 62b46eb0b0bef360c33b591d1dbceb93446701e2

File: Sources_v3/Models/ChannelReadModel.swift (hunk 6 above)
Comment (b-onc) [reply Sources_v3/Models/ChannelReadModel.swift:7]: Added in 62b46eb0b0bef360c33b591d1dbceb93446701e2

File: Sources_v3/APIClient/Endpoints/ChannelEndpoints.swift (hunk 7 above)
Comment (b-onc) [reply Sources_v3/APIClient/Endpoints/ChannelEndpoints.swift:119]: Fixed in d88845270f2dfe90ce3f887da7d256f728e1fb0e

File: Sources_v3/Controllers/ChannelController/ChannelController.swift (hunk 8 above)
Comment (b-onc) [reply Sources_v3/Controllers/ChannelController/ChannelController.swift:588]: Fixed in d88845270f2dfe90ce3f887da7d256f728e1fb0e

File: Sources_v3/APIClient/Endpoints/ChannelEndpoints.swift (hunk 9 above)
Comment (b-onc) [reply Sources_v3/APIClient/Endpoints/ChannelEndpoints.swift:129]: Fixed in 00fda7371027455896dc395d5e012eca690580ef

File: Sources_v3/Controllers/ChannelListController/ChannelListController.swift (hunk 10 above)
Comment (b-onc) [reply Sources_v3/Controllers/ChannelListController/ChannelListController.swift:169]: Fixed in 00fda7371027455896dc395d5e012eca690580ef
There's no nee for guard for new channels, since this endpoint will mark all channels existing at the point of calling as read

Comment (VojtaStavik) [reply Sources_v3/Controllers/ChannelListController/ChannelListController.swift:169]: Indeed 👍

File: Sources_v3/Controllers/ChannelController/ChannelController.swift (hunk 8 above)
Comment (VojtaStavik) [reply Sources_v3/Controllers/ChannelController/ChannelController.swift]: tip: You need to commit first and then mention the hash, otherwise GitHub won't make a link out of the hash.

**Review by VojtaStavik — APPROVED** [review]
Looks great, thanks! 🚢 🇮🇹
